File: Synthetic/nonlinear/py/scenarios.py
"""
scenarios.py — Python counterpart to R/scenarios.R.

Keep these in lock-step: 8 scenarios, 100 reps each, total 800 array tasks.
"""

from itertools import product

# Order must match expand.grid in scenarios.R: N varies fastest, then n_X, then sigma2_x
_N_LIST       = [1000, 4000]
_NX_LIST      = [10, 200]
_SIGMA2X_LIST = [1, 3]

# expand.grid varies the FIRST column fastest, so the outer order is
# (sigma2_x outermost, then n_X, then N innermost)
SCENARIOS = []
sid = 1
for sx in _SIGMA2X_LIST:
    for nx in _NX_LIST:
        for n in _N_LIST:
            SCENARIOS.append({
                "scenario_id": sid,
                "N":           n,
                "n_X":         nx,
                "sigma2_x":    sx,
                "name":        f"N{n}_nX{nx}_sx{sx}",
            })
            sid += 1

N_REPS_PER_SCENARIO = 100


def task_id_to_indices(task_id: int):
    assert 1 <= task_id <= len(SCENARIOS) * N_REPS_PER_SCENARIO, \
        f"task_id {task_id} out of range"
    scenario_idx = (task_id - 1) // N_REPS_PER_SCENARIO        # 0-based
    rep          = (task_id - 1) %  N_REPS_PER_SCENARIO + 1    # 1-based
    s = SCENARIOS[scenario_idx]
    return {
        "task_id":     task_id,
        "scenario_id": s["scenario_id"],
        "rep":         rep,
        "N":           s["N"],
        "n_X":         s["n_X"],
        "sigma2_x":    s["sigma2_x"],
        "name":        s["name"],
    }

File: Synthetic/nonlinear/R/scenarios.R
# =============================================================================
# scenarios.R — Single source of truth for the 8 scenarios.
# Sourced by every R and Python script via the same indexing.
# =============================================================================
#
# Scenario grid:
#   N        ∈ {1000, 4000}
#   n_X      ∈ {10, 200}
#   sigma_x  ∈ {0.7, 2}        # SD of the proxy noise eps_ij ~ N(0, sigma_x^2)
# => 2 * 2 * 2 = 8 scenarios
#
# (sigma_m = sigma_y = 1 are fixed and not part of the grid.)
#
# Reps per scenario: 100
# Total array tasks per method: 8 * 100 = 800
#
# Mapping from SLURM array task id (1..800) to (scenario, rep):
#   scenario_id = ((task_id - 1) %/% 100) + 1   # 1..8
#   rep         = ((task_id - 1) %%  100) + 1   # 1..100
#
# Scenario directory names: N<N>_nX<n_X>_sx<sigma_x>, e.g.
#   N1000_nX10_sx0.7   N4000_nX200_sx2
# =============================================================================

SCENARIOS <- expand.grid(
  N        = c(1000, 4000),
  n_X      = c(10, 200),
  sigma_x  = c(0.7, 2),
  KEEP.OUT.ATTRS = FALSE,
  stringsAsFactors = FALSE
)
SCENARIOS$scenario_id <- seq_len(nrow(SCENARIOS))
# %g (not %d) so that non-integer sigma_x values such as 0.7 format correctly
SCENARIOS$name <- with(SCENARIOS,
                      sprintf("N%d_nX%d_sx%g", N, n_X, sigma_x))

# Fixed noise levels for the mediator and outcome equations (SDs)
SIGMA_M <- 1
SIGMA_Y <- 1

N_REPS_PER_SCENARIO <- 100L

task_id_to_indices <- function(task_id) {
  stopifnot(task_id >= 1, task_id <= nrow(SCENARIOS) * N_REPS_PER_SCENARIO)
  scenario_id <- ((task_id - 1L) %/% N_REPS_PER_SCENARIO) + 1L
  rep         <- ((task_id - 1L) %%  N_REPS_PER_SCENARIO) + 1L
  list(
    task_id     = task_id,
    scenario_id = scenario_id,
    rep         = rep,
    N           = SCENARIOS$N[scenario_id],
    n_X         = SCENARIOS$n_X[scenario_id],
    sigma_x     = SCENARIOS$sigma_x[scenario_id],
    name        = SCENARIOS$name[scenario_id]
  )
}

# Deterministic seed for each (scenario, rep) combo so EVERY method uses the
# same dataset for the same (scenario, rep).
make_seed <- function(scenario_id, rep) {
  as.integer(1000L * scenario_id + rep)
}
